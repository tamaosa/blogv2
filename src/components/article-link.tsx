import React from "react"
import { Link, graphql } from "gatsby"
import { css } from "@emotion/react"

import { ArticleDate } from "./article-date"
import Tags from "./tags"
import { scale, rhythm } from "../utils/typography"

type Props = {
  slug: string
  title: string
  published: string
  updated?: string
  tags?: string[]
}

const titleStyle = css`
  font-size: ${scale(3 / 8).fontSize};
  font-weight: bold;
  line-height: 1;
  a {
    color: var(--fg-title);
    text-decoration: none;
  }
`

const subTitleStyle = css`
  display: flex;
  flex-flow: row wrap;
  align-items: center;
  font-size: ${scale(-1 / 4).fontSize};
  margin: ${rhythm(1 / 4)} 0 ${rhythm(1 / 2)} 0;
`

export const ArticleLink: React.FC<Props> = ({
  slug,
  title,
  published,
  updated,
  tags,
}) => {
  return (
    <article>
      <div css={titleStyle}>
        <Link to={slug} aria-label="記事へのリンク">
          {title}
        </Link>
      </div>
      <div css={subTitleStyle}>
        <div>
          <ArticleDate published={published} updated={updated} />
        </div>
        <div>
          <Tags tags={tags} />
        </div>
      </div>
    </article>
  )
}

export const query = graphql`
  fragment ArticleLink on Mdx {
    fields {
      slug
    }
    frontmatter {
      title
      published
      updated
      tags
    }
  }
`
