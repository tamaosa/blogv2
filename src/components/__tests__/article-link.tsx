import React from "react"
import { render } from "@testing-library/react"
import { ArticleLink } from "../article-link"

describe("ArticleLink", () => {
  it("snapshot test1", () => {
    const { asFragment } = render(
      <ArticleLink slug="/slug" title="tile" published="2022-01-01" />
    )
    expect(asFragment()).toMatchSnapshot()
  })

  it("snapshot test2", () => {
    const { asFragment } = render(
      <ArticleLink
        slug="/slug"
        title="tile"
        published="2022-01-01"
        updated="2022-12-31"
      />
    )
    expect(asFragment()).toMatchSnapshot()
  })

  it("snapshot test3", () => {
    const { asFragment } = render(
      <ArticleLink
        slug="/slug"
        title="tile"
        published="2022-01-01"
        updated="2022-12-31"
        tags={["tag1", "tag2", "tag3", "tag4", "tag5"]}
      />
    )
    expect(asFragment()).toMatchSnapshot()
  })

  it("matching title", () => {
    const { container } = render(
      <ArticleLink slug="/slug" title="tile" published="2022-01-01" />
    )
    expect(container).toHaveTextContent("tile")
  })

  it("matching publication date", () => {
    const { container } = render(
      <ArticleLink slug="/slug" title="tile" published="2022-01-01" />
    )
    expect(container).toHaveTextContent("2022年1月1日 公開")
  })

  it("matching update date", () => {
    const { container } = render(
      <ArticleLink
        slug="/slug"
        title="tile"
        published="2022-01-01"
        updated="2022-12-31"
      />
    )
    expect(container).toHaveTextContent("2022年12月31日 更新")
  })

  it("matching tags", () => {
    const { container } = render(
      <ArticleLink
        slug="/slug"
        title="tile"
        published="2022-01-01"
        updated="2022-12-31"
        tags={["tag1", "tag2", "tag3", "tag4", "tag5"]}
      />
    )
    expect(container).toHaveTextContent("#tag1#tag2#tag3#tag4#tag5")
  })

  it("testing link to a article", () => {
    const slug = "/slug"
    const title = "title"
    const { container } = render(
      <ArticleLink slug={slug} title={title} published="2022-01-01" />
    )
    expect(container.innerHTML).toMatch(
      `<a aria-label="記事へのリンク" href="${slug}">${title}</a>`
    )
  })

  it("testing links to tags", () => {
    const tags = ["tag1", "tag2", "tag3", "tag4", "tag5"]
    const { container } = render(
      <ArticleLink
        slug="/slug"
        title="tile"
        published="2022-01-01"
        updated="2022-12-31"
        tags={tags}
      />
    )
    for (let tag of tags) {
      expect(container.innerHTML).toMatch(
        `<a aria-label="タグへのリンク" href="/tags/${tag}/">#${tag}</a>`
      )
    }
  })
})
