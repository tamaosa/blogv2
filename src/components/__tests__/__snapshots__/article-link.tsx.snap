// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArticleLink snapshot test1 1`] = `
<DocumentFragment>
  .emotion-0 {
  font-size: 1.29684rem;
  font-weight: bold;
  line-height: 1;
}

.emotion-0 a {
  color: var(--fg-title);
  -webkit-text-decoration: none;
  text-decoration: none;
}

.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-flex-flow: row wrap;
  -webkit-flex-flow: row wrap;
  -ms-flex-flow: row wrap;
  flex-flow: row wrap;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-size: 0.8409rem;
  margin: 0.375rem 0 0.75rem 0;
}

.emotion-2 {
  display: inline-block;
  margin-right: 0.375rem;
}

.emotion-3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-flex-flow: row wrap;
  -webkit-flex-flow: row wrap;
  -ms-flex-flow: row wrap;
  flex-flow: row wrap;
  list-style: none;
  margin: 0;
}

.emotion-3 li {
  margin: 0 0.375rem 0 0;
}

.emotion-3 a {
  color: var(--fg);
  border: solid 2px var(--fg);
  border-radius: 10px;
  padding: 0.1875rem;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.emotion-3 a:hover {
  color: var(--bg);
  background: var(--fg-title);
  border: solid 2px var(--fg-title);
}

<article>
    <div
      class="emotion-0"
    >
      <a
        aria-label="記事へのリンク"
        href="/slug"
      >
        tile
      </a>
    </div>
    <div
      class="emotion-1"
    >
      <div>
        <span
          class="emotion-2"
        >
          <time>
            2022年1月1日
          </time>
           公開
        </span>
      </div>
      <div>
        <div>
          <ul
            class="emotion-3"
          />
        </div>
      </div>
    </div>
  </article>
</DocumentFragment>
`;

exports[`ArticleLink snapshot test2 1`] = `
<DocumentFragment>
  .emotion-0 {
  font-size: 1.29684rem;
  font-weight: bold;
  line-height: 1;
}

.emotion-0 a {
  color: var(--fg-title);
  -webkit-text-decoration: none;
  text-decoration: none;
}

.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-flex-flow: row wrap;
  -webkit-flex-flow: row wrap;
  -ms-flex-flow: row wrap;
  flex-flow: row wrap;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-size: 0.8409rem;
  margin: 0.375rem 0 0.75rem 0;
}

.emotion-2 {
  display: inline-block;
  margin-right: 0.375rem;
}

.emotion-4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-flex-flow: row wrap;
  -webkit-flex-flow: row wrap;
  -ms-flex-flow: row wrap;
  flex-flow: row wrap;
  list-style: none;
  margin: 0;
}

.emotion-4 li {
  margin: 0 0.375rem 0 0;
}

.emotion-4 a {
  color: var(--fg);
  border: solid 2px var(--fg);
  border-radius: 10px;
  padding: 0.1875rem;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.emotion-4 a:hover {
  color: var(--bg);
  background: var(--fg-title);
  border: solid 2px var(--fg-title);
}

<article>
    <div
      class="emotion-0"
    >
      <a
        aria-label="記事へのリンク"
        href="/slug"
      >
        tile
      </a>
    </div>
    <div
      class="emotion-1"
    >
      <div>
        <span
          class="emotion-2"
        >
          <time>
            2022年1月1日
          </time>
           公開
        </span>
        <span
          class="emotion-2"
        >
           
          <time>
            2022年12月31日
          </time>
           更新
        </span>
      </div>
      <div>
        <div>
          <ul
            class="emotion-4"
          />
        </div>
      </div>
    </div>
  </article>
</DocumentFragment>
`;

exports[`ArticleLink snapshot test3 1`] = `
<DocumentFragment>
  .emotion-0 {
  font-size: 1.29684rem;
  font-weight: bold;
  line-height: 1;
}

.emotion-0 a {
  color: var(--fg-title);
  -webkit-text-decoration: none;
  text-decoration: none;
}

.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-flex-flow: row wrap;
  -webkit-flex-flow: row wrap;
  -ms-flex-flow: row wrap;
  flex-flow: row wrap;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-size: 0.8409rem;
  margin: 0.375rem 0 0.75rem 0;
}

.emotion-2 {
  display: inline-block;
  margin-right: 0.375rem;
}

.emotion-4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-flex-flow: row wrap;
  -webkit-flex-flow: row wrap;
  -ms-flex-flow: row wrap;
  flex-flow: row wrap;
  list-style: none;
  margin: 0;
}

.emotion-4 li {
  margin: 0 0.375rem 0 0;
}

.emotion-4 a {
  color: var(--fg);
  border: solid 2px var(--fg);
  border-radius: 10px;
  padding: 0.1875rem;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.emotion-4 a:hover {
  color: var(--bg);
  background: var(--fg-title);
  border: solid 2px var(--fg-title);
}

<article>
    <div
      class="emotion-0"
    >
      <a
        aria-label="記事へのリンク"
        href="/slug"
      >
        tile
      </a>
    </div>
    <div
      class="emotion-1"
    >
      <div>
        <span
          class="emotion-2"
        >
          <time>
            2022年1月1日
          </time>
           公開
        </span>
        <span
          class="emotion-2"
        >
           
          <time>
            2022年12月31日
          </time>
           更新
        </span>
      </div>
      <div>
        <div>
          <ul
            class="emotion-4"
          >
            <li>
              <a
                aria-label="タグへのリンク"
                href="/tags/tag1/"
              >
                #tag1
              </a>
            </li>
            <li>
              <a
                aria-label="タグへのリンク"
                href="/tags/tag2/"
              >
                #tag2
              </a>
            </li>
            <li>
              <a
                aria-label="タグへのリンク"
                href="/tags/tag3/"
              >
                #tag3
              </a>
            </li>
            <li>
              <a
                aria-label="タグへのリンク"
                href="/tags/tag4/"
              >
                #tag4
              </a>
            </li>
            <li>
              <a
                aria-label="タグへのリンク"
                href="/tags/tag5/"
              >
                #tag5
              </a>
            </li>
          </ul>
        </div>
      </div>
    </div>
  </article>
</DocumentFragment>
`;
